Mathlib/CategoryTheory/Bicategory/LocallyGroupoid.lean

```diff
@@ -0,0 +1,190 @@
+`Core` to each hom-category of `C`. By construction, `Pith B` is a (2,1)-category,
+and for every (2,1)-category B', every pseudofunctor `B' ⥤ B` factors (essentially) uniquely
+through the inclusion from `Pith B` to `B` (see
+`CategoryTheory.Bicategory.Pith.pseudofunctorToPith`).
+
+## References
+- [Kerodon, section 1.2.2](https://kerodon.net/tag/02GD).
+
+-/
+
+namespace CategoryTheory.Bicategory
+
+open Bicategory
+
+universe w₁ w₂ v₁ v₂ u₁ u₂
+
+/-- A bicategory is locally groupoidal if the categories of 1-morphisms are groupoids. -/
+@[kerodon 009Q]
+abbrev IsLocallyGroupoid (B : Type*) [Bicategory B] := ∀ (b c : B), IsGroupoid (b ⟶ c)
+
+/-- Given a bicategory `B`, `Pith B` is the bicategory obtain by discarding the non-invertible
+2-cells from `B`. We implement this as a wrapper type for `B`, and use `CategoryTheory.Core`
+to discard the non-invertible morphisms. -/
+@[kerodon 00AL]
+structure Pith (B : Type*) where
+  /-- The underlying object of the bicategory. -/
+  as : B
+
+namespace Pith
+
+variable (B : Type u₁)
+
+theorem mk_as (b : Pith B) : mk b.as = b := rfl
+
+instance [Inhabited B] : Inhabited (Pith B) := ⟨⟨default⟩⟩
+
+instance categoryStruct [Bicategory.{w₁, v₁} B] : CategoryStruct (Pith B) where
+  Hom a b := Core (a.as ⟶ b.as)
+  id a := ⟨𝟙 a.as⟩
+  comp f g := ⟨f.of ≫ g.of⟩
+
+variable [Bicategory.{w₁, v₁} B]
+
+-- @[simps!] in categoryStruct puts `Core (a.as ⟶ b.as)` in the hyps for the next two
+-- lemmas, so we record them manually instead.
+@[simp]
+lemma id_of (a : Pith B) : (𝟙 a : a ⟶ a).of = 𝟙 a.as := rfl
+
+@[simp]
+lemma comp_of {a b c : Pith B} (f : a ⟶ b) (g : b ⟶ c) : (f ≫ g).of = f.of ≫ g.of := rfl
+
+instance homGroupoid  (a b : Pith B) :
+    Groupoid.{w₁} (a ⟶ b) := inferInstanceAs <| Groupoid <| Core _
+
+@[ext]
+lemma hom₂_ext {a b : Pith B} {x y : a ⟶ b} {f g : x ⟶ y} {h : f.iso.hom = g.iso.hom} :
+    f = g := CoreHom.ext <| Iso.ext h
+
+@[simp, reassoc]
+lemma comp₂_iso_hom {a b : Pith B} {x y z : a ⟶ b} {f : x ⟶ y} {g : y ⟶ z} :
+    (f ≫ g).iso.hom = f.iso.hom ≫ g.iso.hom := rfl
+
+@[simp, reassoc]
+lemma comp₂_iso_inv {a b : Pith B} {x y z : a ⟶ b} {f : x ⟶ y} {g : y ⟶ z} :
+    (f ≫ g).iso.inv = g.iso.inv ≫ f.iso.inv := rfl
+
+@[simp]
+lemma id₂_iso_hom {a b : Pith B} {x : a ⟶ b} : (𝟙 x : x ⟶ x).iso.hom = 𝟙 _ := rfl
+
+@[simp]
+lemma id₂_iso_inv {a b : Pith B} {x : a ⟶ b} : (𝟙 x : x ⟶ x).iso.inv = 𝟙 _ := rfl
+
+@[simps! whiskerLeft_iso_hom whiskerLeft_iso_inv whiskerRight_iso_hom whiskerRight_iso_inv
+associator_hom_iso associator_inv_iso_hom associator_inv_iso_inv leftUnitor_hom_iso
+leftUnitor_inv_iso_hom rightUnitor_hom_iso rightUnitor_inv_iso_hom rightUnitor_inv_iso_inv]
+instance : Bicategory.{w₁, v₁} (Pith B) where
+  whiskerLeft x _ _ f := CoreHom.mk <| whiskerLeftIso x.of (CoreHom.iso f)
+  whiskerRight f y := CoreHom.mk <| whiskerRightIso (CoreHom.iso f) y.of
+  leftUnitor x :=
+    { hom := CoreHom.mk <| leftUnitor x.of
+      inv := CoreHom.mk (leftUnitor x.of|>.symm) }
```

Comment:
Is it possible to add a definition `Core.isoMk`?

Reply:
It is so possible that it’s now done.